### PR TITLE
Updates to fix android builds failing in Expo

### LINKED ISSUE #546: Gradle build failed for the Expo Android Fix Pre-release
### What were you trying to do?

Trying to install Vision Camera on android and do a successful expo build. 

### Reproduceable Code

_No response_

### What happened instead?

FAILURE: Build failed with an exception.

### Relevant log output

`[stderr] FAILURE: Build failed with an exception. [stderr] * What went wrong: [stderr] Execution failed for task ':app:mergeDebugNativeLibs'. [stderr] > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade [stderr]    > More than one file was found with OS independent path 'lib/x86_64/libreactnativejni.so'. If you are using jniLibs and CMake IMPORTED targets, see https://developer.android.com/studio/preview/features#automatic_packaging_of_prebuilt_dependencies_used_by_cmake [stderr] * Try: [stderr] Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights. [stderr] * Get more help at https://help.gradle.org [stderr] BUILD FAILED in 7m 29s🐛`

### Device

iPhone 12 iOS 14

### VisionCamera Version

2.9.4@expo

### Additional information

- [X] I am using Expo
- [X] I have read the [Troubleshooting Guide](https://mrousavy.github.io/react-native-vision-camera/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

### PR DESCRIPTION
**Pull request by haibert**
## What

This PR applies all the changes from the following [comment](https://github.com/mrousavy/react-native-vision-camera/issues/546#issuecomment-1000849122) which did fix the android build issue when using expo custom-dev-client. 

## Changes

-  in /android/CMakeLists.txt added ${NODE_MODULES_DIR}/react-native/ReactCommon/jsi/jsi/jsi.cpp  in the "add_library" list
- in /android/CMakeLists.txt removed ${JSI_LIB} from the target_link_libraries list
- in /android/build.gradle added the following excludes to the packagingOptions  "**/libreactnativejni.so", "**/libfolly_json.so", "**/libreanimated.so", "**/libjscexecutor.so"
- in grade.properties changed VisionCamera_compileSdkVersion from 31 to 30

⚠️NOTE⚠️
Even with these changes applied the user will need to use the following config plugin (Courtesy of [AdamMercy](https://github.com/mercyaj)) to get a successful build.

```Javascript
const { withProjectBuildGradle } = require('@expo/config-plugins')

const setCompileSdkVersion = (buildGradle, sdkVersion) => {
    const regexpCompileSdkVersion = /\bcompileSdkVersion\s*=\s*(\d+)/
    const match = buildGradle.match(regexpCompileSdkVersion)

    if (match) {
        const existingVersion = parseInt(match[1], 10)

        if (existingVersion < sdkVersion) {
            buildGradle = buildGradle.replace(
                /\bcompileSdkVersion\s*=\s*\d+/,
                `compileSdkVersion = ${sdkVersion}`
            )
        } else {
            throw new Error(`minSdkVersion is already >= ${sdkVersion}`)
        }
    }

    return buildGradle
}

module.exports = (config, sdkVersion) => {
    return withProjectBuildGradle(config, (config) => {
        if (config.modResults.language === 'groovy') {
            config.modResults.contents = setCompileSdkVersion(
                config.modResults.contents,
                sdkVersion
            )
        } else {
            throw new Error(
                "Can't set minSdkVersion in the project build.gradle, because it's not groovy"
            )
        }
        return config
    })
}
```
Then inside of app.json

//---------------IN APP.JSON-----------------
```JSON
"plugins": [
            "react-native-vision-camera",
            ["./withCompileSdkVersion", 31]
        ]
```
//---------------IN APP.JSON-----------------

## Tested on
    * Samsung Galaxy Note8

## Related issues
    * Closes #546

### CODE REVIEW

File: android/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ add_library(
         src/main/cpp/java-bindings/JFrameProcessorPlugin.cpp
         src/main/cpp/java-bindings/JImageProxy.cpp
         src/main/cpp/java-bindings/JHashMap.cpp
+        ${NODE_MODULES_DIR}/react-native/ReactCommon/jsi/jsi/jsi.cpp
```
Comment (mrousavy) [inline android/CMakeLists.txt:25]: This should be avoided if possible. Does it really not build if that line is removed? If not, then please make this conditional (`if` + `set`), so it is not compiled in RN versions that support the `JSI_LIB` library below.

Comment (mrousavy) [reply android/CMakeLists.txt:25]: aka: On RN 0.64 and below, compile `jsi.cpp` here. On RN 0.65 and above, do NOT compile `jsi.cpp` here, but find the `JSI_LIB` below.

File: android/CMakeLists.txt [hunk 2]
```diff
@@ -87,12 +88,6 @@ find_library(
         PATHS ${LIBRN_DIR}
         NO_CMAKE_FIND_ROOT_PATH
 )
-find_library(
-        JSI_LIB
-        jsi
-        PATHS ${LIBRN_DIR}
-        NO_CMAKE_FIND_ROOT_PATH
-)
```
Comment (mrousavy) [inline android/CMakeLists.txt:95]: conditional

File: android/CMakeLists.txt [hunk 3]
```diff
@@ -125,7 +120,6 @@ message(WARNING "VisionCamera linking: FOR_HERMES=${FOR_HERMES}")
 target_link_libraries(
         ${PACKAGE_NAME}
         ${LOG_LIB}
-        ${JSI_LIB}
```
Comment (mrousavy) [inline android/CMakeLists.txt:128]: conditional

File: android/CMakeLists.txt (hunk 1 above)
Comment (mrousavy) [reply android/CMakeLists.txt:25]: the reason for that is, that on RN 0.65 and above, `jsi.cpp` is compiled multiple times - aka we have duplicate symbols. This is a problem in C++, since JSI now exists more than once. My `jsi::JSError` for example is not the same as `jsi::JSError` in React Native, therefore the app hard-crashes. So we need to make that conditional. Check out Reanimated codebase on how they do it

**Review by mrousavy — CHANGES_REQUESTED** [review]
@haibert thank you for the PR bro! 🙏 

While that makes it build for Expo (RN 0.64?), it has some problems in current RN versions. We need to make it conditional, fall back to compiling jsi.cpp if RN version is older. Then it works for both.

**Comment by hirbod** [comment]
And afaik this will break Hermes, right?

**Comment by mrousavy** [comment]
Why would it break Hermes?

**Comment by hirbod** [comment]
@mrousavy that was the latest info I got from @haibert regarding this comment.
https://github.com/mrousavy/react-native-vision-camera/issues/546#issuecomment-1000952046

Haven't verified it yet

**Comment by mrousavy** [comment]
Hey! @haibert @Hirbod can you test if the latest VisionCamera version now successfully builds and runs for you? I finally had the time to investigate those boring build issues, hope everything's resolved for you guys now 🙏

**Comment by hirbod** [comment]
@mrousavy will test this in 30 minutes and report back

**Comment by hirbod** [comment]
@mrousavy I need to dig around this issue first, looks like monorepos aren't still fully supported

```
FAILURE: Build failed with an exception.

* Where:
Build file '/myapp/packages/expo/node_modules/react-native-vision-camera/android/build.gradle' line: 8

* What went wrong:
A problem occurred evaluating project ':react-native-vision-camera'.
> /myapp/packages/expo/node_modules/react-native-vision-camera/node_modules/react-native/ReactAndroid/gradle.properties (No such file or directory)
```

**Comment by hirbod** [comment]
@mrousavy it is not working unfortunately.  (installed 2.11.0)

1) I patched the monorepo path for now, which got rid of the first issue, but then I had to use a config-plugin to bump compileSdkVersion from 30 to 31, also not a big deal, but still a step.

2) After that, I get tons of this here. Gosh I hate android from the bottom of my heart

```log
5): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (82, 20): Class 'kotlin.coroutines.CoroutineContext' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/coroutines/CoroutineContext.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (83, 7): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (90, 4): Class 'kotlin.Suppress' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Suppress.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (93, 20): Class 'kotlin.coroutines.CoroutineContext' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/coroutines/CoroutineContext.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (94, 7): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (104, 20): Class 'kotlin.coroutines.CoroutineContext' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/coroutines/CoroutineContext.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (106, 7): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (107, 14): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (110, 9): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (113, 9): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (120, 5): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (122, 12): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (129, 20): Class 'kotlin.coroutines.CoroutineContext' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/coroutines/CoroutineContext.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (130, 7): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (132, 14): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (143, 20): Class 'kotlin.coroutines.CoroutineContext' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/coroutines/CoroutineContext.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (144, 7): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (154, 30): Unresolved reference: forEach
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (162, 26): Unresolved reference: contains
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (170, 26): Unresolved reference: contains
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (171, 49): Unresolved reference: contains
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (191, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (192, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (193, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (194, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (195, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (196, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (197, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
The class is loaded from /Users/hirbod/.gradle/caches/transforms-3/f824ddc606dcba394bb492d21c1ac821/transformed/jetified-kotlin-stdlib-1.6.0.jar!/kotlin/Unit.class
e: /myapp/node_modules/react-native-vision-camera/android/src/main/java/com/mrousavy/camera/CameraViewModule.kt: (198, 15): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
```

I truncated the output, since its just complaining about the kotlin versions.

**Comment by hirbod** [comment]
@mrousavy alright, after some digging, I found that I had a config-plugin, which inserted following:

```
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.4.10"
    }
```

I had to patch the value from 1.4.10 to 1.5.10 (because 1.6.0 would break expo-dev-launcher) and it looked like that everything will finally compile, but ended up with this error in the last 1% :D

```log
Execution failed for task ':app:mergeDebugNativeLibs'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > More than one file was found with OS independent path 'lib/x86/libc++_shared.so'. If you are using jniLibs and CMake IMPORTED targets, see https://developer.android.com/studio/preview/features#automatic_packaging_of_prebuilt_dependencies_used_by_cmake
```

**Comment by mrousavy** [comment]
Okay thanks for the detailed feedback @Hirbod!

1. Ok for monorepo I will need to figure the node_modules path out clearly.
2. compileSdkVersion still needs to be 31, I don't really want to add that to my config plugin tho, don't think that should be handled on my side...
3. Kotlin version you already resolved
4. Duplicate lib: I'll take a look

**Comment by mrousavy** [comment]
You can also fix that duplicate lib error yourself in your build.gradle btw :) 
That's the recommended approach anyways, a pickFirst is always better than an exclude!

**Comment by hirbod** [comment]
@mrousavy I agree with everything you said. We should just state a little info for the expo users, that they might need a config-plugin which bumps 30 to 31 and add some docs for it.

Regarding pickFirst etc: not exactly sure where to do that + it has to be patched in build.gradle, which would need a config-plugin again. Because that ` excludes = ["**/libc++_shared.so",` is already in your build.gradle

**Comment by haibert** [comment]
@mrousavy Hi Marc! You are the man! 
I had a failed build with the following error

``` Log
[stderr] FAILURE: Build failed with an exception.
[stderr] * Where:
[stderr] Build file '/root/workingdir/build/node_modules/react-native-vision-camera/android/build.gradle' line: 8
[stderr] * What went wrong:
[stderr] A problem occurred evaluating project ':react-native-vision-camera'.
[stderr] > /root/workingdir/build/node_modules/react-native-vision-camera/node_modules/react-native/ReactAndroid/gradle.properties (No such file or directory)
[stderr] * Try:
[stderr] Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
[stderr] * Get more help at https://help.gradle.org
[stderr] BUILD FAILED in 2m 9s
```

**Comment by hirbod** [comment]
@haibert are using a monorepo like me? I had to patch-package like 6 paths to get rid of this. (node_modules/rnvc/android/build.gradle)

But the first error also occured on the first install, because the gradle just got updated after the 2nd build

**Comment by mrousavy** [comment]
where's your node_modules?

**Comment by haibert** [comment]
@mrousavy



its in the root of my project directory

**Comment by hirbod** [comment]
@haibert just try ```expo prebuild --clean```. As said, line 8 was missing for me on the first try and just got updated after the second. I had to dig further though, since I am using a monorepo and all the hardcoded paths have been wrong for me.

**Comment by haibert** [comment]
@Hirbod Hi Hirbod,

I am using eas build, I run the following command "eas build --profile development --platform android" 
I dont think theres a way to do that with eas?

**Comment by hirbod** [comment]
@haibert ok, I see. Well, I will end up there eventually, when my dev process is done and I need to create a release.

@mrousavy I monkey patched my android/app/build.gradle and added every duplicate that was being reported like so:
```
    packagingOptions {
        excludes = ['lib/x86/libc++_shared.so', 'lib/arm64-v8a/libc++_shared.so', 'lib/x86_64/libc++_shared.so', 'lib/armeabi-v7a/libc++_shared.so']
        exclude "META-INF/**"
    }
```

My app finally was building and the development client booted, but ended up like so:

Bildschirmfoto 2022-01-02 um 21 33 21

**Comment by haibert** [comment]
@Hirbod @mrousavy prebuilding didnt give any errors. I guess the error is with when you use eas build to create the a build from the cloud? maybe the nodemodules isnt in the root directory when you send your project to the cloud im guessing.. Have you tried to do a release build Hirbod?

**Comment by hirbod** [comment]
@mrousavy alright, while I do not consider this a solution, FOR NOW, my app is building (but still crashing due to reanimated and dev tools).

I had to patch it like so:

```
    packagingOptions {
        pickFirst 'lib/x86/libc++_shared.so'
        pickFirst 'lib/arm64-v8a/libc++_shared.so'
        pickFirst 'lib/x86_64/libc++_shared.so'
        pickFirst 'lib/armeabi-v7a/libc++_shared.so'
    }
```

@haibert no, I never used "eas" cloud stuff so far.


EDIT: IT IS FINALLY WORKING. As said, with the patch, but I have a running android app, cant believe it!

**Comment by mrousavy** [comment]
@Hirbod seems like you figured it out already.

`exclude` actually drops the libraries and doesn't link them. `pickFirst` uses the first of those available. So with your first "monkeypatch" you stole a required library from your app 😄 

The `pickFirst` approach seems reasonable, after all this is because React Native doesn't offer good support for native library consumers (prefabs). I also have this in my regular React Native app. It works because the libc++ library version is the same anyways. 

So in conclusion, what needs to be done is:

1. Fix the monorepo path finding for node_modules.
2. Wait until React Native ships their native libraries as prefabs, then you don't need `pickFirst` anymore.

**Comment by hirbod** [comment]
Hi Marc,

thanks for the clarification. So I would say we need to provide / update the config plug-in and add the above to /android/app/build.gradle

This could be a workaround. Shall we open an issue upstream or is this something already known?

Edit:
@mrousavy looks like someone built a approach here:
https://github.com/mrousavy/react-native-vision-camera/pull/545#issuecomment-949477264

**Comment by haibert** [comment]
@mrousavy Hi Marc,

So 2.11.1 got rid of the monorepo issue. I had to use a plugin to set the compileSdkVersion to 31, but I got the build error bellow 

``` Log
FAILURE: Build failed with an exception.
[stderr] * What went wrong:
[stderr] Execution failed for task ':app:mergeDebugNativeLibs'.
[stderr] > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
[stderr]    > More than one file was found with OS independent path 'lib/x86/libhermes.so'. If you are using jniLibs and CMake IMPORTED targets, see https://developer.android.com/studio/preview/features#automatic_packaging_of_prebuilt_dependencies_used_by_cmake

```

**Comment by hirbod** [comment]
@haibert check out my PR
https://github.com/mrousavy/react-native-vision-camera/pull/708

```
    [
      'react-native-vision-camera',
      {
        enableMicrophonePermission: true,
        disableFrameProcessors: true,
        dangerouslyHandleAndroidSharedLibrary: true
      }
    ],
```

I've added two new options for android (disable frame proceccsors, if you don't need them + fixing the sharedlibrary issue that you also encountered).

Thanks for your report, this helped me to change the glob pattern.

**Comment by haibert** [comment]
@Hirbod awesome! Thank you I will try this out!

**Comment by hirbod** [comment]
@haibert my PR will now be able to disable frame processors also for iOS.

**Comment by hirbod** [comment]
This PR is obsolete.

Since my dangerouslyHandleAndroidSharedLibrary prop was removed, people will most likely encounter issues with expo-dev-client.

So it is necessary to use  the minSdkVersion plugin (bump to 31) and the pickFirst config plugin. (see below)


```ts
const {
    withAppBuildGradle
} = require('@expo/config-plugins')

const addPickFirst = (buildGradle, paths) => {
    const regexpPackagingOptions = /\bpackagingOptions\s*{[^}]*}/
    const packagingOptionsMatch = buildGradle.match(regexpPackagingOptions)

    const bodyLines = []

    paths.forEach((path) => bodyLines.push(`pickFirst '${path}'`))

    const body = bodyLines.join('\n        ')

    if (packagingOptionsMatch) {
        console.warn(
            'WARN: withPickFirst: Replacing packagingOptions in app build.gradle'
        )

        return buildGradle.replace(
            regexpPackagingOptions,
            `packagingOptions {
        ${body}
       }
      `
        )
    }

    const regexpAndroid = /\nandroid\s*{/
    const androidMatch = buildGradle.match(regexpAndroid)

    if (androidMatch) {
        return buildGradle.replace(
            regexpAndroid,
            `
android {
    packagingOptions {
        ${body}
    }
     `
        )
    }

    throw new Error('withPickFirst: Could not find where to add packagingOptions')
}

module.exports = (config, paths) => {
    if (!paths) {
        throw new Error('withPickFirst: No paths specified!')
    }

    return withAppBuildGradle(config, (config) => {
        if (config.modResults.language === 'groovy') {
            config.modResults.contents = addPickFirst(
                config.modResults.contents,
                paths
            )
        } else {
            throw new Error(
                "withPickFirst: Can't add pickFirst(s) because app build.grandle is not groovy"
            )
        }
        return config
    })
}
```

and install it like this

```js
    ['./plugins/withPickFirst', ['lib/x86/libc++_shared.so', 'lib/x86_64/libc++_shared.so', 'lib/armeabi-v7a/libc++_shared.so', 'lib/arm64-v8a/libc++_shared.so']],
```

**Comment by mrousavy** [comment]
Thanks for the explanation @Hirbod!
